### DISCUSSION TITLE
Support for SPM package projects in fastlane scan

### DISCUSSION THREAD

**Discussion by nostradani** [fastlane/fastlane#17362] [Ideas]
### Summary

Currently it seems that you cannot use fastlane scan for pure SPM package projects i.e. projects that only use a Package.swift and no Xcode project. 

### Reasoning

When developing SPM packages for iOS,  you cannot use "swift build" or "swift test" to compile or test your package. Instead you have to use xcodebuild. Therefore also fastlane spm cannot be used.
For building a package "fastlane xcodebuild" can be used, but with regards to testing the user is left on his own (e.g. no pretty test results as html etc.)

### Current status

fastlane scan expects either a Xcode workspace or project file, which it validates upfront. Therefore there is no way clean way to get fastlane scan to work with only a Package.swift

However xcodebuild itself works as intended if only a Package.swift is present.

### Proposal

- Introduce a new boolean setting like "use_spm" to indicate the use with a SPM package
- Adjust the validation to check for an existing Package.swift file
- The remaining behaviour should remain the same

### Related issues

We were able to circumvent the aforementioned restrictions by using a custom script instead of the xcodebuild command. There we noticed some "command timed out" messages which seemed to happen while xcodebuild resolved other SPM dependencies. We already know these from older fastlane versions in conjunction with fastlane gym. Our current fastlane version have them resolved in gym, but it still seems to persist in fastlane scan.

We are currently using fastlane  2.162.0

**Comment by H2wk** [comment]
I would like to build a little on this as I have found all the similar issues. The `xcodeproj: ` parameter is mandatory in Fastlane at the moment.

When building a swift package on the command line with `xcodebuild` you can build & test the package without a `xcodeproj` or 
`xcworkspace`.

Xcode automatically generates a `scheme` file for any swift package manager projects so there is no longer a need to have a `xcodeproj` or `xcworkspace` file present. 

**To test:** 

1. Create a new swift package. `swift package init`. The swift package will be named after the folder you created it in. 
2. Run `xcodebuild -list` this will list the schemes associated with the project. When this command is complete you will be able to see the scheme exists in the project and is named according to the directory it is in. 
3. Build the project: xcodebuild -scheme %NAME% build`
4. Test the project: `xcodebuild -scheme %NAME% test`

Would be great to see this functionality added where the project or workspace can be specified as optional. 

Link to the mandatory fields for scan [here](https://docs.fastlane.tools/actions/scan/#parameters) 

Same issue with [xcov](https://github.com/fastlane-community/xcov/issues/190)

**Comment by macblazer** [comment]
This is also a problem for any Swift 5.3+ packages that use xib resources.  `swift build` and `swift test` do not compile the xib into a nib, or place it in it's correct location on disk (within the module's bundle `Contents/Resources` folder).  `xcodebuild test` does handle xib resources in swift packages correctly.  See this long discussion for [resource build nuances](https://forums.swift.org/t/swift-5-3-spm-resources-in-tests-uses-wrong-bundle-path/37051/23).

Allowing Fastlane scan to work without an xcodeproj but requiring a scheme if a Package.swift is present would solve this problem.

**Comment by macblazer** [comment]
Issue #17734 is the one to watch for the `scan` tool specifically.

**Comment by CAMOBAP** [comment]
I was managed to build a package for iOS with some `-Xswiftc` flags like: 

`swift build -Xswiftc "-sdk" -Xswiftc "`xcrun --sdk iphonesimulator --show-sdk-path`" -Xswiftc "-target" -Xswiftc "x86_64-apple-ios9.0-simulator"`

it will be nice if `spm` action will support some property like this:

```
spm (
   ...
   xswiftc: ['-sdk', sdkp_path, '-target', 'x86_64-apple-ios9.0-simulator']
   ...
)
```

**Comment by Cyrus113** [comment]
Thank you for sharing this post i had the issue which i got solved thanks buddy..
[https://www.myloweslife.ltd/](url)

**Comment by adamdahan** [comment]
Is there any actual documentation on how to make fastlane scan work with standalone swift package? Like tutorial style? I can't make any sense from the docs.fastlane.tools regarding such a thing. I see this merged: `Issue https://github.com/fastlane/fastlane/issues/17734 is the one to watch for the scan tool specifically.` but 0 information on how to use it?

**Comment by macblazer** [comment]
This is more or less the minimum set of options needed to run tests on a Swift package.  This assumes that the Package.swift is in the same folder as your fastlane folder, and your package is named "YourPackageName".

```
params = { package_path: '.', scheme: 'YourPackageName' }
run_tests(params)
```

If your package supports iOS/tvOS/iPadOS/watchOS you can supply the `devices` option to have it run tests on one or more device simulators (or actual devices if they are connected):

```
params = { package_path: '.', scheme: 'YourPackageName', devices: ['iPhone 11 Pro'] }
```

**Comment by tmaly1980** [comment]
They've finally removed generate-xcodeproj from the "swift package"  command in Xcode 14.3.1. Meaning we cannot generate project files anymore for swift packages. How do we get around this? 

There is a "swift test" command but it expects to run for macos too and I'm not supporting that. There doesn't seem to be a way to build/test just for iOS platforms.

**Comment by tmaly1980** [comment]
I did a bit of research on how to run tests from the CLI using xcodebuild without a project/workspace file and also looked under the hood with how scan generates the xcodebuild command and came to this conclusion as to how to automate unit tests (on CircleCI) for a pure Swift package:

```
platform :ios do
  desc "Run Tests"
  lane :test do
    scan(
      package_path: ".", # This makes swift package testing possible without a pbxproj or xcworkspace file
      scheme: "PACKAGENAME-Package", # Needs "-Package" at the end 
      clean: true,
      device: "iPhone 11 Pro", # This is needed for a proper destination
      output_types: 'junit',
      output_files: 'junit.xml',
      result_bundle: true, # Other errors with xcoutput required me to add this
      output_directory: Dir.pwd + "/test_output"
    )
  end
end
```

Works like a charm!

**Reply by khanboy1989** [reply]
what if you would like to build it first and then run the test script how would you do it? 

def build_lane
	scan(scheme: ENV["FASTLANE_SCHEME"],
		devices: ["iPhone 14", "iPad Pro (12.9-inch) (6th generation)"],
		clean: true,
		force_quit_simulator: true,
		prelaunch_simulator: true,
		build_for_testing: true,
		package_path: '.',
		derived_data_path: ENV["DERIVED_DATA_TEST"],
		ensure_devices_found: true,
		result_bundle: true,
		output_directory: Dir.pwd + "/test_output",
    		code_coverage: true,
	)
end

I am having issue with my build lines, it does fail sometime and it gives me an error like this:
xcodebuild: error: Failed writing xctestrun file: The folder “PackageName_iphonesimulator17.2-arm64.xctestrun” doesn’t exist..

**Comment by Foerster64** [comment]
As of my last update in January 2022, Fastlane's `scan` tool did not directly support Swift Package Manager (SPM) projects. However, you can still use `scan` to run tests for SPM projects with a few extra steps. Here's how:

1. Use xcodebuild with SPM: Since SPM projects are built using `xcodebuild`, you can use `xcodebuild` commands to run tests for your SPM project. For example, you can use the following command to run tests for an SPM project:
   ```
   xcodebuild test -scheme YourSPMScheme -destination 'platform=iOS Simulator,name=iPhone 12'
   ```

2. Integrate with Fastlane: You can integrate `xcodebuild` commands into your Fastfile by using the `sh` action. Here's an example of how you can do this.[.](https://valerocreditcard.fun/)...: 
   ```ruby
   lane :test_spm_project do
     sh "xcodebuild test -scheme YourSPMScheme -destination 'platform=iOS Simulator,name=iPhone 12'"
   end
   ```

3. Run Tests using Scan: Alternatively, you can use `scan` along with `xcodebuild` commands within a lane. For example:
   ```ruby
   lane :test_spm_project do
     sh "xcodebuild build-for-testing -scheme YourSPMScheme -destination 'platform=iOS Simulator,name=iPhone 12'"
     scan(scheme: "YourSPMScheme", clean: true)
   end
   ```

Remember to replace `"YourSPMScheme"` with the actual scheme name of your SPM project, and adjust the destination accordingly.

While `scan` doesn't have built-in support for SPM projects, you can still leverage `xcodebuild` commands within Fastlane to achieve similar functionality. Additionally, always check the latest updates and documentation for Fastlane to see if there have been any changes or additions related to SPM project support.
